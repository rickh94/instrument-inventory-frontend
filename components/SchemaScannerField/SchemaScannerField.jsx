import React, { useState } from 'react'
import PropTypes from 'prop-types'

import connectField from 'uniforms/connectField'
import {
  FormControl,
  InputLabel,
  Button,
  Input,
  InputAdornment,
  FormHelperText,
} from '@material-ui/core'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBarcode } from '@fortawesome/free-solid-svg-icons'

import { Scanner, TooltipIconButton } from '..'

export const SchemaScannerField = ({ onChange, value, error, label, required }) => {
  const autoScan = JSON.parse(localStorage.getItem('autoScan', 'false'))
  const [scanning, setScanning] = useState(autoScan)

  const onDetected = result => {
    if (result.codeResult.code !== value) {
      setScanning(false)
      onChange(result.codeResult.code)
    }
  }

  return (
    <FormControl fullWidth error={error ? true : false}>
      <InputLabel htmlFor="scanner-field" required={required}>
        {label}
      </InputLabel>
      {scanning ? (
        <React.Fragment>
          <Scanner onDetected={onDetected} />
          <Button onClick={() => setScanning(false)} data-testid="stop-scanning-button">
            Stop Scanning
          </Button>
        </React.Fragment>
      ) : (
        <React.Fragment>
          <Input
            id="scanner-field"
            onChange={event => onChange(event.target.value)}
            aria-describedby="scanner-error"
            type="text"
            value={value}
            required={required}
            data-testid="scanner-input"
            endAdornment={
              <InputAdornment position="end">
                <TooltipIconButton
                  onClick={() => setScanning(true)}
                  title="Scan Barcode"
                  data-testid="scan-button"
                >
                  <FontAwesomeIcon icon={faBarcode} />
                </TooltipIconButton>
              </InputAdornment>
            }
          />
          {error && <FormHelperText id="scanner-error">{error.message}</FormHelperText>}
        </React.Fragment>
      )}
    </FormControl>
  )
}

SchemaScannerField.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  error: PropTypes.object,
  label: PropTypes.string.isRequired,
  required: PropTypes.bool,
}

export default connectField(SchemaScannerField)
