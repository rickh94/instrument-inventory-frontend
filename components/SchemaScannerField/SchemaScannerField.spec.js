import React from 'react'
import { SchemaScannerField } from './SchemaScannerField'
import { render } from '@testing-library/react'

describe('<SchemaScannerField />', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <SchemaScannerField onChange={jest.fn()} value="" error={null} label="Test" />
    )
    expect(container).toMatchSnapshot()
  })
})
