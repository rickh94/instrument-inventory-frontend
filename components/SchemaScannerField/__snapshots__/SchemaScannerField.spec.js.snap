// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SchemaScannerField /> matches snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="scanner-field"
    >
      Test
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
    >
      <input
        aria-describedby="scanner-error"
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
        id="scanner-field"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          data-testid="scan-button"
          tabindex="0"
          title="Scan Barcode"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-barcode fa-w-16 "
              data-icon="barcode"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 448V64h18v384H0zm26.857-.273V64H36v383.727h-9.143zm27.143 0V64h8.857v383.727H54zm44.857 0V64h8.857v383.727h-8.857zm36 0V64h17.714v383.727h-17.714zm44.857 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm35.715 0V64h18v383.727h-18zm44.857 0V64h18v383.727h-18zm35.999 0V64h18.001v383.727h-18.001zm36.001 0V64h18.001v383.727h-18.001zm26.857 0V64h18v383.727h-18zm45.143 0V64h26.857v383.727h-26.857zm35.714 0V64h9.143v383.727H476zm18 .273V64h18v384h-18z"
                fill="currentColor"
              />
            </svg>
            <span
              class="MuiTypography-root MuiTypography-srOnly"
              data-testid="sr-text"
            >
              Scan Barcode
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
